llvm/lib/CodeGen/RegAllocFast.cpp

```diff
@@ -1368,6 +1371,40 @@ void RegAllocFastImpl::findAndSortDefOperandIndexes(const MachineInstr &MI) {
   });
 }
 
+/// Compute \ref UseOperandIndexes so it contains the indices of "use" operands
+/// that are to be allocated. Those are ordered in a way that high-priority
+/// classes are allocated first.
+void RegAllocFastImpl::findAndSortUseOperandIndexes(const MachineInstr &MI) {
```

Comment:
I'd expect this function to mirror the def case, and use the size of the classes

Reply:
I agree that it would be better for consistency, I'll add that code back in.

However, I think we still need to use `AllocationPriority` to fix #98222, because the class sizes heuristic cannot spot that two register classes contain overlapping (but not identical) registers. There's a FIXME comment in `addRegClassDefCounts` to take that into account, but I don't think that `TargetRegisterInfo` contains the information we'd need for that.

Reply:
This is actually a big problem for AMDGPU. There are a number of heuristics around that assume the number of registers in a class are distinctly allocatable, and does not consider overlapping cases. We really need a separate query for the number of distinct registers in the class. The priority is still a workaround for not having this information, and requires this subtle maintenance burden of keeping class priorities ordered

Reply:
I've had another look through `TargetRegisterInfo` but I still don't think it has the information about the overlap between register classes which we need here. Are you OK with using the `AllocationPriority` as a workaround for now?